demo-circuit/src/naive_threshold_sig_w_key_rotation/mod.rs

```diff
@@ -0,0 +1,789 @@
+            cs.ns(|| "genesis_constant == validators_keys_root if prev cert is missing"),
+            &genesis_validator_keys_tree_root_g,
+            &is_first_certificate_g,
+        )?;
+
+        // 2B) Additional check: Check validator key root and threshold hash
+        // Enforce H(validator key root, threshold) to be equal to H(genesis_validator_keys_tree_root_g, threshold)
+        // or H(all_leaves_root_g, threshold) if it is the first certificate
+
+        //in the case of the first certificate, check if expected_vkroot_threshold_hash_g equals to evaluated vkroot_threshold_hash_g with initial set of keys
+        //otherwise, in case of new certificates, use the genesis_validator_keys_tree_root
+        let mut vkroot_threshold_hash_g = FieldElementGadget::conditionally_select(
+            cs.ns(|| "select initial keys for threshold check"), 
+            &is_first_certificate_g, 
+            &validator_mktree_leaves_root_g, 
+            &genesis_validator_keys_tree_root_g,
+        )?;
+
+        vkroot_threshold_hash_g = FieldHashGadget::enforce_hash_constant_length(
+            cs.ns(|| "H(H(pks), threshold)"),
+            &[vkroot_threshold_hash_g, threshold_g.clone()],
+        )?;
+
+        let expected_vkroot_threshold_hash_g =
+            FieldElementGadget::alloc_input(cs.ns(|| "alloc constant as input"), || {
+                self.genesis_constant
+                    .ok_or(SynthesisError::Other("AssignmentMissing genesis constant".to_string()))
+            })?;
+
+            vkroot_threshold_hash_g.enforce_equal(
+            cs.ns(|| "vkroot_threshold_hash: expected == actual"),
+            &expected_vkroot_threshold_hash_g,
+        )?;
+
+
+        // ****************** 3 Verify wcert signatures ********************
+        // for (i <- 0 to (VALIDATORS_SIZE-1)) {
+        //     var sigNum = 0                  // we will count number of collected signatures
+        //     if (wcert_signatures[i] != NULL) {
+        //         val is_valid_sig = verify_signature(sc_wcert_hash, wcert_signatures[i], signing_keys[i])      // order of signatures should corresponed the order of keys in signing_keys
+        //         require(is_valid_sig == true)
+        //         sigNum = sigNum + 1    
+        //     }
+        // }
+        // require(sigNum > VALIDATORS_SIZE/2)  CR: This is replaced by the threshold value
+   
+        //Alloc the b's as witnesses
+        let mut bs_g = Vec::with_capacity(log_max_pks + 1);
+        for (i, b) in self.b.iter().enumerate() {
+            let b_g = Boolean::alloc(cs.ns(|| format!("alloc b_{}", i)), || {
+                b.ok_or(SynthesisError::Other(format!("AssignmentMissing b {}", i)))
+            })?;
+            bs_g.push(b_g);
+        }
+
+        //Pack the b's into a field element
+        let b_field = FieldElementGadget::from_bits(
+            cs.ns(|| "pack the b's into a field element"),
+            bs_g.as_slice(),
+        )?;
```

Comment:
Let's organize the code better please if possible, as now the checks are completely mixed